minimal_scheduling/minimal_sched_fifo_callback_group.cpp

```diff
@@ -0,0 +1,151 @@
+  {
+    return realtime_callback_group_;
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer1_, timer2_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher1_, publisher2_;
+  size_t count1_, count2_;
+  rclcpp::CallbackGroup::SharedPtr realtime_callback_group_;
+};
+
+class MinimalSubscriber : public rclcpp::Node
+{
+public:
+  MinimalSubscriber()
+  : Node("minimal_subscriber")
+  {
+    subscription1_ = this->create_subscription<std_msgs::msg::String>(
+      "topic",
+      10,
+      [this](std_msgs::msg::String::UniquePtr msg) {
+        RCLCPP_INFO(this->get_logger(), "I heard:        '%s'", msg->data.c_str());
+      });
+
+    realtime_callback_group_ = this->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+    rclcpp::SubscriptionOptions subscription_options;
+    subscription_options.callback_group = realtime_callback_group_;
+    subscription2_ = this->create_subscription<std_msgs::msg::String>(
+      "topic_rt",
+      10,
+      [this](std_msgs::msg::String::UniquePtr msg) {
+        RCLCPP_INFO(this->get_logger(), "I heard (RT):   '%s'", msg->data.c_str());
+      }, subscription_options);
+  }
+
+  rclcpp::CallbackGroup::SharedPtr get_realtime_callback_group()
+  {
+    return realtime_callback_group_;
+  }
+
+private:
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription1_;
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription2_;
+  rclcpp::CallbackGroup::SharedPtr realtime_callback_group_;
+};
+
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  auto node_pub = std::make_shared<MinimalPublisher>();
+  auto node_sub = std::make_shared<MinimalSubscriber>();
+
+  rclcpp::executors::StaticSingleThreadedExecutor default_callback_group_executor;
+  rclcpp::executors::StaticSingleThreadedExecutor realtime_executor;
+
+  default_callback_group_executor.add_node(node_pub);
+  default_callback_group_executor.add_node(node_sub);
+
+  // the real-time callback groups are added to a specific executor which is run in a thread
+  // with real-time scheduling
+  // note we could have more than one executor with different scheduling policies and priorities
+  realtime_executor.add_callback_group(
+    node_pub->get_realtime_callback_group(), node_pub->get_node_base_interface());
+  realtime_executor.add_callback_group(
+    node_sub->get_realtime_callback_group(), node_sub->get_node_base_interface());
+
+  // spin real-time tasks in a separate thread
+  auto realtime_thread = std::thread(
+    [&]() {
+      realtime_executor.spin();
+    });
+
+  struct sched_param param {};
+  int policy = SCHED_FIFO;
+  param.sched_priority = 90;
+  auto ret = pthread_setschedparam(realtime_thread.native_handle(), policy, &param);
+  if (ret > 0) {
+    std::cerr << "Couldn't set scheduling priority and policy. Error code " << strerror(errno) << std::endl;
+    return EXIT_FAILURE;
+  }
```

Comment:
Minor nitpick: do you want to refactor this stuff into a `create_rt_thread(std::function)` method that creates the std::thread and then set the RT priority + policy? Might make the code show the intention a bit better?

Reply:
Actually, I like the fact that we show how to configure a given `std::thread` by getting the native handle.